packages/wrangler/src/experimental-assets.ts

```diff
@@ -1,8 +1,299 @@
+							e
+						)
+					);
+					// Exponential backoff, 1 second first time, then 2 second, then 4 second etc.
+					await new Promise((resolvePromise) =>
+						setTimeout(resolvePromise, Math.pow(2, attempts) * 1000)
+					);
+					if (e instanceof APIError && e.isGatewayError()) {
+						// Gateway problem, wait for some additional time and set concurrency to 1
+						queue.concurrency = 1;
+						await new Promise((resolvePromise) =>
+							setTimeout(resolvePromise, Math.pow(2, gatewayErrors) * 5000)
+						);
+						gatewayErrors++;
+						// only count as a failed attempt after a few initial gateway errors
+						if (gatewayErrors >= MAX_UPLOAD_GATEWAY_ERRORS) {
+							attempts++;
+						}
+					} else {
+						attempts++;
+					}
+					return doUpload();
+				} else if (isJwtExpired(initializeAssetsResponse.jwt)) {
+					throw new FatalError(
+						`Asset upload took too long on bucket ${bucketIndex + 1}/${initializeAssetsResponse.buckets.length}. Please try again.
+						Assets already uploaded have been saved, so the next attempt will automatically resume from this point.`
+					);
+				} else {
+					throw e;
+				}
+			}
+		};
+		// add to queue and run it if we haven't reached concurrency limit
+		void queue.add(() =>
+			doUpload().then((res) => {
+				completionJwt = res.jwt || completionJwt;
+			})
+		);
+	}
+	queue.on("error", (error) => {
+		logger.error(error.message);
+		throw error;
+	});
+	await queue.onIdle();
+
+	// if queue finishes without receiving JWT from asset upload service (AUS)
+	// AUS only returns this in the final bucket upload response
+	if (!completionJwt) {
+		throw new FatalError(
+			"Failed to complete asset upload. Please try again.",
+			1
+		);
+	}
+
+	const uploadMs = Date.now() - start;
+	const skipped = Object.keys(manifest).length - numberFilesToUpload;
+	const skippedMessage = skipped > 0 ? `(${skipped} already uploaded) ` : "";
+
+	logger.log(
+		`✨ Success! Uploaded ${numberFilesToUpload} file(s) ${skippedMessage}${formatTime(uploadMs)}\n`
+	);
+
+	return completionJwt;
+};
+
+// modified from /pages/validate.tsx
+const walk = async (
+	dir: string,
+	manifest: AssetManifest,
+	startingDir: string = dir
+) => {
+	const files = await readdir(dir);
+
+	let counter = 0;
+	await Promise.all(
+		files.map(async (file) => {
+			const filepath = path.join(dir, file);
+			const relativeFilepath = path.relative(startingDir, filepath);
+			const filestat = await stat(filepath);
+
+			if (filestat.isSymbolicLink()) {
+				return;
+			}
+
+			if (filestat.isDirectory()) {
+				manifest = await walk(filepath, manifest, startingDir);
+			} else {
+				if (counter >= MAX_ASSET_COUNT) {
+					throw new UserError(
+						`Max asset count exceeded: ${counter.toLocaleString()} files found.
+						You cannot have more than ${MAX_ASSET_COUNT.toLocaleString()} files in a deployment.
+						Ensure you have specified your build output directory correctly (currently set as ${startingDir}).`
```

Comment:
> `Ensure you have specified your build output directory correctly (currently set as ${startingDir}).`

the action we are recommending here feels a bit misleading to me. We're pointing the problem at the assets directory, when in fact it's the number of files that's the problem. I expect in most cases the assets directory specififed by users will be the correct one, and the action they need to take is to restrict the number of assets in said directory. To that effect, I wonder if a better action here would be


```suggestion
					throw new UserError(
						`Maximum number of assets exceeded.
						Cloudflare Workers supports up to ${MAX_ASSET_COUNT.toLocaleString()} assets in a deployment. We found ${counter.toLocaleString()} files in the specified assets directory "${startingDir}".
						Ensure your assets directory contains a maximum of ${MAX_ASSET_COUNT.toLocaleString()} files, or that you have specified your assets directory correctly.`
```